confluent_kafka/serialization/avro.py

```diff
@@ -0,0 +1,227 @@
+        :param SerializationContext ctx:
+
+        :raises: SerializerError
+
+        :returns: None
+        :rtype: None
+        """
+        if datum is None:
+            return None
+
+        self._encode(fo, datum, ctx)
+
+    def _encode(self, fo, schema, datum, ctx):
+        """
+        Given a parsed Schema, encode a record for the given topic.  The
+        record is expected to be a dictionary.
+        The schema is registered with the subject of 'topic-value'
+
+        :param SerializationContext ctx: Topic name
+        :param object datum: Avro object to serialize
+        :returns: Encoded record with schema ID as bytes
+        :rtype: bytes
+
+        :raises: SerializerError
+        """
+
+        if schema is None:
+            raise SerializerError("Missing schema.")
+
+        subject = self.name_func(schema, ctx)
+        self.registry.register(subject, schema)
+
+        # Write the magic byte and schema ID in network byte order (big endian)
+        fo.write(struct.pack('>bI', MAGIC_BYTE, schema.id))
+        # write the record to the rest of the buffer
+        schemaless_writer(fo, schema.schema, datum)
+
+        return
+
+    def deserialize(self, data, ctx):
+        """
+        Decode Avro Binary to Avro object
+
+        :param bytes data: Avro binary
+        :param SerializationContext ctx: Optional schema to project object on.
+
+        :raises SerializerError
+
+        :returns: decoded Avro object
+        :rtype: object
+        """
+        if data is None:
+            return None
+
+        record = self._decode(data, self.reader_schema)
+
+        if self.converter:
+            return self.converter.from_dict(record)
+
+        return record
+
+    def _decode(self, data, reader_schema=None):
+        """
+        Decode a message from kafka that has been encoded for use with
+        the schema registry.
+
+        :param str|bytes or None data: message key or value to be decoded
+        :param Schema reader_schema: Optional compatible Schema to project value on.
+
+        :returns: Decoded value.
+        :rtype object:
+        """
+
+        if len(data) <= 5:
+            raise SerializerError("message is too small to decode {}".format(data))
+
+        with ContextStringIO(data) as payload:
+            magic, schema_id = struct.unpack('>bI', payload.read(5))
+            if magic != MAGIC_BYTE:
+                raise SerializerError("message does not start with magic byte")
+
+            schema = self.registry.get_by_id(schema_id)
+
+            if reader_schema:
+                return schemaless_reader(payload, schema.schema, reader_schema.schema)
+            return schemaless_reader(payload, schema.schema)
+
+
+class Converter(object):
+    """
+    py.class:: Converter
+
+    Converts class instance objects to and from dict instances to enable more sophisticated
+    Avro record serialization/deserialization.
+
+    """
+    def to_dict(self, record_obj):
```

Comment:
having this as a default implementation seems like it would do more harm than good to me. same with from_dict.

Reply:
agreed